MDEV-12302: Execute triggers for foreign key updates/deletes

Simoffsky opened the pull request:
- [x] *The Jira issue number for this PR is: MDEV-12302*


## Description
In this approach, an sql-layer api is added for calling on the row delete or update. These callbacks are supposed to be expandable for various needs, and ideally should be capable of everything that would happen during usual update/delete row events: check constraints, binlogging, etc, and in particular, trigger invocation, which is implemented in the base commit.

Innodb fills up the sql-format records and invokes the sql-side. It re-uses an equsting update_node, serving for cascade action, and delivers que_thr_t for re-use. As a result, cascade calls are provisioned with the information on call depth and can handle the depth exceeded errors as before.

For correct error handling of whatever could happen on the sql side, a new error code is introduced both for handler api and innodb: HA_ERR_CASCADE_SQL aka DB_SQL_ERROR. This error is distinguished from generic DB_ERROR: the latter is supposed for the unclassifiable cases and shoudn't normally be exposed through the handler api (so it's a backup plan code).

Foreign key depth control is further refectored for better error handling, more efficient and clear code.

Roadmap:
✅ Simple Delete
✅ Simple Update
✅ Fetch data from secondary indexes
✅ Handle changes coming from BEFORE triggers
✅ Virtual fields
✅ Test ON DELETE/UPDATE SET NULL
✅  Bit fields (probably just test it)
✅ Long blobs

## Release Notes
Added support for triggers on events from cascade delete/update foreign key actions. The triggers are invoked the same way as for usual delete/update/insert events.

<!--
Tick one of the following boxes [x] to help us understand if the base branch for the PR is correct.
-->
## Basing the PR against the correct MariaDB version
- [x] *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
- [ ] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

<!--
  All code merged into the MariaDB codebase must meet a quality standard and codying style.
  Maintainers are happy to point out inconsistencies but in order to speed up the review and merge process we ask you to check the CODING standards.
-->
## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=3496) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=3496) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=3496) it.</sub>

FooBarrior commented:
Hello @Simoffsky! As the project is close to its final stage, it's time to reorganize your code, also covering @vuvova's request from #3354.

Please move the triggers processing (and other sql-layer "hooks") code to the sql layer.

Let's consider two approaches to this.

## 1. make two sql-side functions related to hooks: before and after
I will refer to hooks as to ephemere trigger like actions, that are to be done before or after a row change.

We can already observe that before hooks differ from after hooks: before hooks also include updating virtual fields, whereas after hooks don't. In the close future, the list will expand (like with row logging, CHEK constraints, etc).

So it makes sense for making two functions.

This approach would look as placing two functions in `sql_class.cc`:
```
int sql_do_before_triggers(TABLE*, bool is_delete);
int sql_do_after_triggers(TABLE*, bool is_delete);
```

This also could make sense to be done in **four** functions instead of two.

Then, in `ha_innodb.cc`, adding an sql-format row conversion function, which also retreives the necessary sql-layer objects, like maria_table:
```
dberr_t innodb_convert_ib_row_to_sql(dict_index_t *, row_t *,..., TABLE **maria_table /*!< out */,  ...);
```

This approach also requires a manual conversion back from sql format (in case if triggers change the data)
```
dberr_t innodb_convert_sql_row_to_ib(...);
```

## 2. Call an sql-layer function instead of innodb's `row_upd`.

We can instead make a whole row processing on the sql side:
`sql_table.cc` would be a suitable place for the following two functions:
```
sql_update_row(TABLE *);
sql_delete_row(TABLE *);
```

One will have to call `table->file->ha_update_row` (or `ha_delete_row`) inside.

Also positioning the cursor with `table->file->ha_rnd_pos` is mandatory before any `ha_update_row` and `ha_delete_row` calls. 

In `ha_innodb.cc`, adding
```
dberr_t innodb_do_foreign_cascade(dict_table_t *, upd_node_t*,...);
```
and replace the `row_upd_step` call in `row_update_cascade_for_mysql`.

## Which way to choose

1 preserves innodb row_upd_step call, which makes things faster, especially if no hooks are required to call.

OTOH, 2 gives a more flexibility on what the sql layer can do. We can also replace ha_delete_row with `TABLE::delete_row` call, which can eventyally bring System Versioning support in the sql layer. This is actually very desired.

2 also needs no manual conversion backward, which can be less error-prone, and just easier.

sql_update_row and sql_delete_row wil allso be able to be eventually re-used `Sql_cmd_{update|delete}`, or in other commands, that have to make general-purpose row updates (like `REPLACE` or `INSERT...ON DUPLICATE KEY UPDATE`, or yet unimplemented `MERGE`)

So I'm asking you to go on with approach 2.

FooBarrior commented:
Please also join all tests in a single file. Test execution time is a problem for us, as every test run has a significant latency.

Note that we maintain 80 characters width for the c/c++ sode, with putting an opening `{` brace on a blank line.
Though https://github.com/MariaDB/server/blob/main/CODING_STANDARDS.md holds just a basic set of rules, it's better to read and follow it to pass a review with less edits.

Simoffsky committed:
MDEV-12302 Execute triggers for foreign key updates/deletes

FooBarrior reviewed (commented):
@Simoffsky I'm going to have a time to focus on my leftovers, and this patch is of my high interest. 

Let's go through it, perhaps, and see what could we miss?

FooBarrior committed:
MDEV-12302 Add SQL errors handling

Bypass the errors from sql layer, through innodb back to sql layer.
New handler error: HA_ERR_CASCADE_SQL
New innodb error: DB_SQL_ERROR

This error is not visible to user and serves only for correct error
transport. The real error code is expected to be stored in
thd->m_stmt_da, that is, already reported

This is a fixup commit.
Original author: Simoffsky <maks-kirsanov26@mail.ru>

FooBarrior committed:
MDEV-12302 Refactor fk cascade depth check

* Move inc and dec for fk_cascade_depth close to each other, in a
symmetrical manner, wrapping the recursive call.

* Remove conputation-heavy node traversal with
row_ins_cascade_n_ancestors for each cascade call,
resulting in O(depth^2) total passes. It should be enough to do that
only as an assertion. Historically, there could be more than one thr,
joined in chains (but then we'd better pass over a chain of thr). Now a
single thr is used for cascade call, so no need in any graph traversal.

* Fix max cascade error handling -- now HA_ERR_FK_DEPTH_EXCEEDED is
returned, as implied.

FooBarrior committed:
MDEV-12302 Move fk cascade depth control to THD

que_thr_t is not bypassed in between trigger calls, that is the count
starts from 0 each time in the trigger.

This means that an effective depth restriction is a multiplication of
FK_MAX_CASCADE_DEL and max_sp_recursion_depth.

Moving the control to THD changes this to addition.

Now the depth limit is: FK_MAX_CASCADE_DEL + max_sp_recursion_depth

FooBarrior committed:
Fix versioning.foreign crash

The problem is row_update_vers_insert invoked before the cascade update.
Cascade update relies on pre-defined node->pcur, and an insert can
invalidate the cursor, if b-tree node split happens.

One possible solution is to move the versioned insert after cascade
update.

This patch goes the described way, but also moves versioned insert to
the sql layer.

We have to re-set row event bits in trg_event_map, because
node->is_delete is determined based on those values.

TODO is to fix versioning.foreign 'trx_id,unique' result mismatch.

FooBarrior committed:
MDEV-31942 Online alter: support cascade foreign keys

All the technical provisioning was done yet by
MDEV-12302 Execute triggers for foreign key updates/deletes

As it executes ha_update_row and ha_delete_row directly.

All that is left is remove the restriction and add the tests.

FooBarrior committed:
Fix cascade operations overwriting the record, when handler is re-used
in the nested row event.

FooBarrior committed:
MDEV-38840 optimize store_record/restore_record to use aligned copy

FooBarrior committed:
MDEV-38840 optimize store_record/restore_record to use aligned copy

benchmarks.

To see the result in realistic load, run:
python run_mariadb_bench_inlined.py > results_final.plain
python convert_benchmarks.py

FooBarrior committed:
MDEV-38840 add avx512 option

midenok commented on sql/unireg.h:
```diff
@@ -71,8 +71,9 @@
 	/* Extern defines */
-#define store_record(A,B) memcpy((A)->B,(A)->record[0],(size_t) (A)->s->reclength)
-#define restore_record(A,B) memcpy((A)->record[0],(A)->B,(size_t) (A)->s->reclength)
+#include "my_memcpy_bench.h"
+#define store_record(A,B) copy_record((A), (A)->B,(A)->record[0])
+#define restore_record(A,B) copy_record((A), (A)->record[0],(A)->B)
```
Please, revert or counterargument.

midenok commented on sql/table.cc:
```diff
@@ -2277,10 +2270,10 @@ int TABLE_SHARE::init_from_binary_frm_image(THD *thd, bool write,
-  rec_buff_length= ALIGN_SIZE(share->reclength + 1);
+  rec_buff_length= MY_ALIGN(share->reclength + 1, RECORD_ALIGNMENT);
   share->rec_buff_length= rec_buff_length;
-  if (!(record= (uchar *) alloc_root(&share->mem_root, rec_buff_length)))
-    goto err;                          /* purecov: inspected */
+  record= record_alloc(&share->mem_root, 1);
+
   /* Mark bytes after record as not accessable to catch overrun bugs */
   MEM_NOACCESS(record + share->reclength, rec_buff_length - share->reclength);
```
Please, fix.

midenok commented on sql/sql_table.cc:
```diff
@@ -13936,3 +13894,117 @@ bool HA_CREATE_INFO::
+    table->update_virtual_fields(table->file, VCOL_UPDATE_FOR_READ);
+  
+  table->column_bitmaps_set(&table->s->all_set, &table->s->all_set);
+  if (table->triggers &&
+      unlikely(table->triggers->process_triggers(thd, TRG_EVENT_DELETE,
+                                                 TRG_ACTION_BEFORE, FALSE,
+                                                 &trg_skip_row)))
+    return trg_skip_row ? HA_ERR_CASCADE_SQL : 0;
```
Please, fix or comment in the code.

midenok commented on sql/sql_table.cc:
```diff
@@ -13936,3 +13894,117 @@ bool HA_CREATE_INFO::
+  // This ensures that triggers can correctly read virtual field values
+  update_virtual_fields_for_rows(table);
+
+  if (table->triggers && 
+        table->triggers->has_triggers(TRG_EVENT_UPDATE, TRG_ACTION_BEFORE)) 
+  {
+      if (unlikely(table->triggers->process_triggers(thd, TRG_EVENT_UPDATE,
+                                                     TRG_ACTION_BEFORE, TRUE,
+                                                     &trg_skip_row)))
+        return trg_skip_row ? 0 : HA_ERR_CASCADE_SQL;
+      // This is necessary for indexes that depend on virtual fields
+      update_virtual_fields_for_rows(table);
+  }
```
Please, fix.

midenok commented on include/my_memcpy_bench.h:
```diff
@@ -0,0 +1,423 @@
+#pragma once
+#include <atomic>
+#include <chrono>
+#include <cstring>
+#include <cassert>
```
Please, fix.

midenok commented on run_mariadb_bench_inlined.py:
```diff
@@ -0,0 +1,195 @@
+
+debug = 0
+
+funcs = [
+    # "memcpy_threshold_8byte_mov_avx_fallback", requires power of 2 alignment
+    "memcpy_manual_avx512_loop_aligned"
+    "memcpy_manual_avx_loop_aligned",
+    "memcpy_manual_avx_loop_unaligned",
+    "memcpy_manual_sse_loop",
+    "memcpy_threshold16",
+    "memcpy_rep_movsq",
+    "memcpy_aligned<8>",
+    "memcpy_aligned<16>",
+    "memcpy_aligned<avx>",  # 32-byte aligned
+    "memcpy",
+    "memcpy_best_aligned",
+    "memcpy_best_unaligned",
+    "my_exact_unaligned_memcpy",
+    "memcpy_switch_inline_aligned",
+    "memcpy_threshold_builtins16",
+    "memcpy_threshold_builtins8",
+    "memcpy_threshold_builtins_avx",
+]
```
Please, fix.

midenok commented on run_mariadb_bench_inlined.py:
```diff
@@ -0,0 +1,195 @@
+                run_cmd(f"./bld_gemini/client/mariadb -u root --socket=/tmp/mariadb_bench.sock bench_db -e \"CREATE TABLE t1 (id INT PRIMARY KEY AUTO_INCREMENT, val VARCHAR({size})) ENGINE={engine} /*WITH SYSTEM VERSIONING*/;\"")
+
+                # Insert data
+                run_cmd(f'./bld_gemini/client/mariadb -u root --socket=/tmp/mariadb_bench.sock bench_db -e "insert t1 select seq, \'{'a'*size}\' from seq_1_to_1000000"')
```
Please, fix.

midenok commented on run_mariadb_bench_inlined.py:
```diff
@@ -0,0 +1,195 @@
+
+        build("mariadbd")
+
+        if not start_mariadbd():
+            print(f"  Server failed to start for {func} {engine}/{size}. Skipping.", flush=True)
+            sys.exit(1)
```
Please, fix.

midenok commented on convert_benchmarks.py:
```diff
@@ -0,0 +1,88 @@
+        "benchmarks": benchmarks
+    }
+
+    # Generate JSON file
+    json_file_path = f'benchmark/results_{engine.lower()}.json'
+    with open(json_file_path, 'w') as f:
+        json.dump(output_json, f, indent=2)
+    print(f"Generated {json_file_path}")
```
Please, fix.

midenok reviewed (changes requested):
Please fix build failures (and make sure the tests pass). Please fix Copilot comments.

gkodinov reviewed (changes requested):
Thank you for your work on this. This is (an attempt at) a preliminary review. 

Please make sure the PR is actually ready for review before you move it to "open". Right now it doesn't look like it is: it doesn't compile and there's a bunch of commits in it that most probably will need to be reorganized and/or squashed. 

I'd suggest keeping this as "draft" until it is in a better shape.
